tests/Regressions.hs

```diff
@@ -71,6 +76,57 @@ propEqAfterDelete (Keys keys) =
 mapFromKeys :: [Int] -> HM.HashMap Int ()
 mapFromKeys keys = HM.fromList (zip keys (repeat ()))
 
+------------------------------------------------------------------------
+-- Issue #254
+
+-- Key type that always collides.
+newtype KC = KC Int
+  deriving (Eq, Ord, Show)
+instance Hashable KC where
+  hashWithSalt salt _ = salt
+
+issue254Lazy :: Assertion
+issue254Lazy = issue254LazyLambda 2
+
+-- We want to make sure that old values in the HashMap are evicted when new values are inserted,
+-- even if they aren't evaluated. To do that, we use the WeakPtr trick described at
+-- http://simonmar.github.io/posts/2018-06-20-Finding-fixing-space-leaks.html.
+-- We insert a value named oldV into the HashMap, then insert over it, checking oldV is no longer reachable.
+--
+-- To make the test robust, it's important that oldV isn't hoisted up to the top or shared. To do that,
+-- we use NOINLINE, make oldV dependent on an unseen argument, and insert _ <- return () to ensure oldV
+-- is under a lambda.
+{-# NOINLINE issue254LazyLambda #-}
+issue254LazyLambda :: Int -> Assertion
+issue254LazyLambda i = do
+  _ <- return () -- put oldV under a lambda
+  let oldV = error $ "Should not be evaluated: " ++ show i
+  weakV <- mkWeakPtr oldV Nothing -- test whether oldV is alive
+  let mp = HML.insert (KC 1) (error "Should not be evaluated") $ HML.fromList [(KC 0, "1"), (KC 1, oldV)]
+  _ <- evaluate mp -- force the insert to happen
+  performGC
+  res <- deRefWeak weakV -- gives Just if oldV is still alive
+  _ <- evaluate mp -- makes sure that we didn't GC away the whole HashMap, just oldV
```

Comment:
Don't `evaluate mp` here; that doesn't really explain our aim, and in principle GHC could simply erase it because `mp` has already been evaluated. What we really want to do is `touch mp`.

```haskell
import GHC.Exts ( touch# )
import GHC.IO (IO (..))
touch :: a -> IO ()
touch a = IO (\s -> (# touch# a s, () #))
```

The very latest GHC version has a new `with#` function (not yet a primop) that's more reliable than `touch#`, but I don't think anything we're doing here could trip the flakiness in `touch#`.

Reply:
My desire to reach into the guts of the IO monad is low - I think GHC really should expose `touch` in Control.Exception itself. But in the absence of that, I've copied the code above.